cpp/src/replace/nulls.cu

```diff
@@ -374,55 +374,41 @@ struct replace_policy_functor {
-    if (replace_policy == cudf::replace_policy::PRECEDING) {
-      thrust::inclusive_scan(
-        rmm::exec_policy(stream), in_begin, in_begin + input.size(), gm_begin, func);
-    } else {
-      auto in_rbegin = thrust::make_reverse_iterator(in_begin + input.size());
-      auto gm_rbegin = thrust::make_reverse_iterator(gm_begin + gather_map.size());
-      thrust::inclusive_scan(
-        rmm::exec_policy(stream), in_rbegin, in_rbegin + input.size(), gm_rbegin, func);
-    }
-
-    auto output = cudf::detail::gather(cudf::table_view({input}),
-                                       gather_map.begin(),
-                                       gather_map.end(),
-                                       cudf::out_of_bounds_policy::DONT_CHECK);
 
-    return std::move(output->release()[0]);
+std::unique_ptr<cudf::column> replace_nulls_policy_impl(cudf::column_view const& input,
+                                                        cudf::replace_policy const& replace_policy,
+                                                        rmm::cuda_stream_view stream,
+                                                        rmm::mr::device_memory_resource* mr)
+{
+  auto device_in = cudf::column_device_view::create(input);
+  auto index     = thrust::make_counting_iterator<cudf::size_type>(0);
+  auto valid_it  = cudf::detail::make_validity_iterator(*device_in);
+  auto in_begin  = thrust::make_zip_iterator(thrust::make_tuple(index, valid_it));
+
+  rmm::device_vector<cudf::size_type> gather_map(input.size());
+  auto gm_begin = thrust::make_zip_iterator(
+    thrust::make_tuple(gather_map.begin(), thrust::make_discard_iterator()));
+
+  auto func = replace_policy_functor();
+  if (replace_policy == cudf::replace_policy::PRECEDING) {
+    thrust::inclusive_scan(
+      rmm::exec_policy(stream), in_begin, in_begin + input.size(), gm_begin, func);
+  } else {
+    auto in_rbegin = thrust::make_reverse_iterator(in_begin + input.size());
+    auto gm_rbegin = thrust::make_reverse_iterator(gm_begin + gather_map.size());
+    thrust::inclusive_scan(
+      rmm::exec_policy(stream), in_rbegin, in_rbegin + input.size(), gm_rbegin, func);
   }
 
-  template <typename col_type, std::enable_if_t<not cudf::is_fixed_width<col_type>()>* = nullptr>
-  std::unique_ptr<cudf::column> operator()(cudf::column_view const& input,
-                                           cudf::replace_policy const& fillna_policy,
-                                           rmm::cuda_stream_view stream,
-                                           rmm::mr::device_memory_resource* mr)
-  {
-    CUDF_FAIL("No specialization exists for the given type.");
-  }
-};
+  auto output = cudf::detail::gather(cudf::table_view({input}),
+                                     gather_map.begin(),
+                                     gather_map.end(),
+                                     cudf::out_of_bounds_policy::DONT_CHECK);
+
+  return std::move(output->release()[0]);
```

Comment:
This is a modern C++ antipattern. `return std::move(...)` actively prohibits copy elision. https://stackoverflow.com/a/17473874/749748

```suggestion
  return output->release()[0];
```

Reply:
So this is a rare situation where you actually do need the `std::move` in the return. Normally we can return a `unique_ptr` stack variable from a function without needing `std::move`.

This situation is different because we're not returning a stack variable. We're returning an element of a `std::vector`, i.e., `table::release()` returns a `std::vector<unique_ptr<column>>`. We need to move ownership of the `unique_ptr<column>` from the vector in order to be able to return it.

Here's a simpler example of the same behavior: https://godbolt.org/z/8beM59

Reply:
Damnit, I hate inconsistent rules. Thanks Jake.

Good to see that in this particular case, at least the version without `move` fails to compile! So can the rule be summarized to `Don't `return std::move` except when the compiler requires it`?

Reply:
Alternatively you could do something like this:
```
std::unique_ptr<int> foo2(){
    std::vector<std::unique_ptr<int>> v;
    v.emplace_back(std::make_unique<int>(42));
    auto p = std::move(v[0]); 
    return p;
}
```
This separates moving the ownership from the vector and returning the object so that you can still apply the rule of saying "Never use `std::move` on a returned object".   

> "Don't return std::move except when the compiler requires it"

Yeah, this is a more accurate rule.